packages/asset-swapper/src/network/samplers/uniswap_v3.ts

```diff
@@ -0,0 +1,122 @@
+import { ChainId } from '@0x/contract-addresses';
+import { BigNumber } from '@0x/utils';
+
+import { UniswapV3BuySamplerContract, UniswapV3SellSamplerContract } from '../../wrappers';
+import { constants } from '../../constants';
+
+import { Chain } from '../chain';
+import { OnChainSourceSampler, UnwrapContractFunctionReturnType, SamplerEthCall } from '../source_sampler';
+import { Address, Bytes, ERC20BridgeSource, FillData } from "../types";
+import { valueByChainId } from '../utils';
+
+const { NULL_ADDRESS } = constants;
+
+interface SamplerConfig {
+    quoter: Address,
+    router: Address,
+};
+
+const UNISWAP_V3_CONFIG_BY_CHAIN_ID = valueByChainId(
+    {
+        [ChainId.Mainnet]: {
+            quoter: '0xb27308f9f90d607463bb33ea1bebb41c27ce5ab6',
+            router: '0xe592427a0aece92de3edee1f18e0157c05861564',
+        },
+        [ChainId.Ropsten]: {
+            quoter: '0x2f9e608fd881861b8916257b76613cb22ee0652c',
+            router: '0x03782388516e94fcd4c18666303601a12aa729ea',
+        },
+    },
+    { quoter: NULL_ADDRESS, router: NULL_ADDRESS },
+);
+
+export interface UniswapV3FillData extends FillData {
+    tokenAddressPath: Address[];
+    router: Address;
+    uniswapPath: Bytes;
+}
+
+type SellContract = UniswapV3SellSamplerContract;
+type BuyContract = UniswapV3BuySamplerContract;
+type SellContractSellFunction = SellContract['sampleSells'];
+type BuyContractBuyFunction = BuyContract['sampleBuys'];
+type SellContractSellFunctionParameters = Parameters<SellContractSellFunction>;
+type BuyContractBuyFunctionParameters = Parameters<BuyContractBuyFunction>;
+type SellContractSellFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<SellContractSellFunction>>;
+type BuyContractBuyFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<BuyContractBuyFunction>>;
+type SamplerSellEthCall = SamplerEthCall<UniswapV3FillData,SellContractSellFunctionParameters,SellContractSellFunctionReturn>;
+type SamplerBuyEthCall = SamplerEthCall<UniswapV3FillData,BuyContractBuyFunctionParameters,BuyContractBuyFunctionReturn>;
```

Comment:
Nuking this boilerplate would be pretty good 😂

Reply:
I tried so hard but ultimately gave up. If only TS had preprocessors.